R/ab_scenario.R

```diff
@@ -18,122 +20,90 @@
+  ...
+) {
+  if(methods::is(od, class2 = "sf")) {
+    od = sf::st_drop_geometry(od)
+  }
+  # minimise n. columns:
+  od = od[c(names(od)[1:2], modes)]
+  od_long = tidyr::pivot_longer(od, cols = modes, names_to = "mode")
+  repeat_indices = rep(seq(nrow(od_long)), od_long$value)
+  od_longer = od_long[repeat_indices, 1:3]
+  # summary(od_longer$geo_code1 %in% zones$geo_code)
+  if(!is.null(subpoints)) {
+    suppressMessages({
+      subpoints = sf::st_centroid(subpoints)
     })
-    # todo: Allow multiple scenarios to be calculated here? (RL 2020-02-10)
-    mode_cname = paste0("mode_", scenario)
-    desire_lines_disag[[mode_cname]] = NA
-    n_walk = desire_lines[[match_scenario_mode(cnames, scenario, mode = "Walk")]][i]
-    n_bike = desire_lines[[match_scenario_mode(cnames, scenario, mode = "Bike|cycle")]][i]
-    n_drive = desire_lines[[match_scenario_mode(cnames, scenario, mode = "car_d|drive")]][i]
-    if(any(grepl(pattern = "transit|bus|rail", x = cnames))) {
-      n_transit = desire_lines[[match_scenario_mode(cnames, scenario, mode = "transit|bus|rail")]][i]
-    } else {
-      n_transit = 0
-    }
-    corrent_n = n_walk + n_bike + n_transit + n_drive == pop
-
-    # fix edge cases where n. people travelling by modes do not match population
-    # todo: update input data (RL)
-    n_mismatch = nrow(desire_lines_disag) -  sum(n_walk, n_bike, n_drive, n_transit)
-    if(n_mismatch != 0) {
-      warning("Mismatch between n. trips and population. Check your data")
-      n_walk = n_walk + n_mismatch
-    }
-
-    # todo: update this next line? (RL)
-    if(n_walk > 0) {
-      desire_lines_disag[[mode_cname]][sample(nrow(desire_lines_disag), size = n_walk)] = "Walk"
-    }
-    no_mode = which(is.na(desire_lines_disag[[mode_cname]]))
-    if(length(no_mode) > 0) {
-      desire_lines_disag[[mode_cname]][sample(no_mode, size = n_bike)] = "Bike"
-    }
-    no_mode = which(is.na(desire_lines_disag[[mode_cname]]))
-    if(length(no_mode) > 0) {
-      desire_lines_disag[[mode_cname]][sample(no_mode, size = n_drive)] = "Drive"
-    }
-    # Other modes include taking public transit and being a passenger in a car. A/B Street doesn't
-    # model the latter, so for now map all of these to transit. Also note that bus routes are mostly
-    # not imported yet, so transit trips will wind up walking.
-    desire_lines_disag[[mode_cname]][is.na(desire_lines_disag[[mode_cname]])] = "Transit"
-    if(i == 1) {
-      desire_lines_out = desire_lines_disag
-    } else {
-      desire_lines_out = rbind(desire_lines_out, desire_lines_disag)
-    }
   }
-  if(output_format == "sf") {
-    return(desire_lines_out)
-  } else {
-    return(ab_json(desire_lines_out, mode_column = mode_cname, time_fun = time_fun, ...))
+  if(!is.null(subpoints_d)) {
+    suppressMessages({
+      subpoints_d = sf::st_centroid(subpoints_d)
+    })
   }
+  od::od_jitter(od = od_longer, z = zones, subpoints = subpoints, subpoints_d = subpoints_d)
```

Comment:
If you pass in an entire polygon for the zones, then above, you use `st_centroid` and collapse that down to a single point. Then here, you jitter all of those centroids. How can `od_jitter` possibly know how much to jitter the centroids, when the polygon has been collapsed? Wouldn't it be conceptually simpler to preserve the full polygon, then later as you need to, sample from points inside of the polygon?

Reply:
`st_centroid` is used to find the centre points of the buildings, not the centrepoints of the zones. By default `od_jitter()` picks random start and end points anywhere in each zone but if you give it pre-selected possible start and end points (e.g. buildings) it will make the trips start and end at these points. Not 100% sure I've answered the question, maybe one to chat about over audio with a screenshare.

Reply:
Ah I misunderstood what it was jittering; this makes sense, thanks. I think matching up the param names in the docs with the code would also resolve this confusion